composer/utils/s3_object_store.py

```diff
@@ -0,0 +1,124 @@
+            extra (Optional[Dict], optional): Extra attributes to pass to the underlying provider driver.
+                (default: ``None``, which is equivalent to an empty dictionary)
+            headers (Optional[Dict[str, str]], optional): Additional request headers, such as CORS headers.
+                (defaults: ``None``, which is equivalent to an empty dictionary)
+        """
+        del headers, verify_hash # used in the libcloud interface
+
+        if bucket is None:
+            if self.bucket is None:
+                raise ValueError("No S3 bucket specified")
+            bucket = self.bucket
+
+        self.client.upload_file(Filename=file_path, Bucket=bucket, Key=object_name, ExtraArgs=extra)
+
+    def upload_object_via_stream(self,
+                                 obj: Union[bytes, Iterator[bytes]],
+                                 object_name: str,
+                                 bucket: Optional[str] = None,
+                                 extra: Optional[Dict] = None,
+                                 headers: Optional[Dict[str, str]] = None):
+        """Upload an object.
+
+        Args:
+            obj (bytes | Iterator[bytes]): The object.
+            object_name (str): Object name (i.e. where the object will be stored in the container.)
+            verify_hash (bool, optional): Whether to verify hashes (default: ``True``)
+            extra (Optional[Dict], optional): Extra attributes to pass to the underlying provider driver.
+                (default: ``None``)
+            headers (Optional[Dict[str, str]], optional): Additional request headers, such as CORS headers.
+                (defaults: ``None``)
+        """
+        del headers
+
+        if bucket is None:
+            if self.bucket is None:
+                raise ValueError("No S3 bucket specified")
+            bucket = self.bucket
+        
+        # add handling for other bytes-like types
+        if not isinstance(obj, bytes):
+            raise NotImplementedError("S3ObjectStore.upload_object_via_stream only takes a bytes object.")
+        
+        obj = io.BytesIO(obj)
+
+        # if isinstance(obj, bytes):
+        #     obj = iter(i.to_bytes(1, sys.byteorder) for i in obj)
+        
+        self.client.upload_fileobj(obj, Bucket=bucket, Key=object_name, ExtraArgs=extra)
+
+    def download_object(self,
+                        object_name: str,
+                        destination_path: str,
+                        bucket: Optional[str] = None,
+                        overwrite_existing: bool = False,
+                        delete_on_failure: bool = True):
+        """Download an object to the specified destination path.
+        """
+
+        if bucket is None:
+            if self.bucket is None:
+                raise ValueError("No S3 bucket specified")
+            bucket = self.bucket
+
+        self.client.download_file(Bucket=bucket, Key=object_name, Filename=destination_path)
```

Comment:
Can `overwrite_existing` be implemented (e.g. error if it is false and the file already exists?)

In addition, can you download to a temporary file path (e.g. `destination_path + f".{time.time()}.tmp"`, and then once the file has been downloaded, `os.rename` it to `destination_path`)? If the download fails, then can we make a best ensure that the tempfile is automatically cleaned up (probably via try/finally?)